python/cuml/metrics/pairwise_distances.pyx

```diff
@@ -136,6 +134,104 @@ def _determine_metric(metric_str, is_sparse=False):
         return PAIRWISE_DISTANCE_METRICS[metric_str]
 
 
+def nan_euclidean_distances(
+    X, Y=None, *, squared=False, missing_values=cp.nan
+):
+    """Calculate the euclidean distances in the presence of missing values.
+
+    Compute the euclidean distance between each pair of samples in X and Y,
+    where Y=X is assumed if Y=None. When calculating the distance between a
+    pair of samples, this formulation ignores feature coordinates with a
+    missing value in either sample and scales up the weight of the remaining
+    coordinates:
+
+        dist(x,y) = sqrt(weight * sq. distance from present coordinates)
+        where,
+        weight = Total # of coordinates / # of present coordinates
+
+    For example, the distance between ``[3, na, na, 6]`` and ``[1, na, 4, 5]``
+    is:
+
+        .. math::
+            \\sqrt{\\frac{4}{2}((3-1)^2 + (6-5)^2)}
+
+    If all the coordinates are missing or if there are no common present
+    coordinates then NaN is returned for that pair.
+
+    Read more in the :ref:`User Guide <metrics>`.
+
+    Parameters
+    ----------
+    X : array-like of shape (n_samples_X, n_features)
+        An array where each row is a sample and each column is a feature.
+
+    Y : array-like of shape (n_samples_Y, n_features), default=None
+        An array where each row is a sample and each column is a feature.
+        If `None`, method uses `Y=X`.
+
+    squared : bool, default=False
+        Return squared Euclidean distances.
+
+    missing_values : np.nan or int, default=np.nan
+        Representation of missing value.
+
+    Returns
+    -------
+    distances : ndarray of shape (n_samples_X, n_samples_Y)
+        Returns the distances between the row vectors of `X`
+        and the row vectors of `Y`.
+    """
+
+    if is_sparse(X):
+        X = X.toarray()
+    if is_sparse(Y):
+        Y = Y.toarray()
+
+    if Y is None:
+        Y = X
+
+    # Get missing mask for X
+    missing_X = _get_mask(X, missing_values)
+
+    # Get missing mask for Y
+    missing_Y = missing_X if Y is X else _get_mask(Y, missing_values)
+
+    # set missing values to zero
+    X[missing_X] = 0
+    Y[missing_Y] = 0
+
+    # Adjust distances for sqaured
+    distances = cp.array(pairwise_distances(X, Y, metric="sqeuclidean"))
+
+    # Adjust distances for missing values
+    XX = X * X
```

Comment:
This seems really expensive- we're essentially copying each input just to compute a masked l2 norm that we can subtract from the pairwise distance matrix. In the case where X == Y, we're copying the input twice. It should be fairly straightforward to do this in place, even using a `RawKernel` if needed. You should be able to take the storage requirement down to `X.shape[0]` (or `X.shape[0]` + `Y.shape[0]` when Y is supplied).